File: tauri-plugin-co-sdk/guest-js/types/co.ts
import { CID } from "multiformats";
import { Network } from "./network.js";

export type CoId = string;
export type Did = string;

export type Tags = string[][];
/**
 * Rust: #[co] pub struct Core
 */
export interface Core {
  /** The CID of the core binary. */
  binary: CID;

  /** Core tags. */
  tags: Tags;

  /** The latest stream state (Option<Cid>). */
  state?: CID;
}

/**
 * Rust: #[co] pub struct Guard
 */
export interface Guard {
  /** The CID of the guard binary. */
  binary: CID;

  /** Guard tags. */
  tags: Tags;
}

export enum ParticipantState {
  Active = "Active",
  Invite = "Invite",
  // add more if needed
}

export type Participant = {
  did: Did;
  state: ParticipantState;
  tags?: Tags;
};

/**
 * Rust: #[co] pub struct Key
 */
export interface Key {
  /** Key ID string. */
  id: string;

  /** Key state (repr(u8) in Rust). */
  state: KeyState;
}

/**
 * Rust: #[co(repr)] #[repr(u8)] pub enum KeyState
 */
export enum KeyState {
  Inactive = 0,
  Active = 1,
}

export interface Co {
  /** CO UUID. */
  id: CoId;

  /** CO Tags. (default, omitted if empty) */
  t?: Tags;

  /** CO Name. */
  n: string;

  /** CO Core Binary (CID). */
  b: CID;

  /** CO Current heads. (BTreeSet<Cid> -> array) */
  heads: CID[];

  /**
   * CO Participants. (CoMap<Did, Participant> -> object map)
   * default, omitted if empty
   */
  p?: CID;

  /**
   * CO Streams with the associated state reference.
   * Key: Core Instance (string)
   * TODO Comes as object but should be map
   */
  c: { [key: string]: Core };

  /** Co Guards. default, omitted if empty */
  g?: Map<string, Guard>;

  /**
   * CO Encryption Keys.
   * The first (index 0) key is the active key.
   * Option<Vec<Key>> -> omitted if null/undefined.
   */
  k?: Key[];

  /**
   * CO network services. default, omitted if empty
   * Assuming CoSet<Network> serializes like a set -> array.
   */
  s?: Network[];
}

export type CoAction =
  | {
      Upgrade: {
        binary: CID;
        migrate?: CID;
      };
    }
  | {
      Heads: {
        heads: Set<CID>;
      };
    }
  | {
      TagsInsert: {
        tags: Tags;
      };
    }
  | {
      TagsRemove: {
        tags: Tags;
      };
    }
  | {
      ParticipantInvite: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      ParticipantJoin: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      ParticipantPending: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      ParticipantRemove: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      ParticipantTagsInsert: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      ParticipantTagsRemove: {
        participant: Did;
        tags: Tags;
      };
    }
  | {
      NetworkInsert: {
        network: Network;
      };
    }
  | {
      NetworkRemove: {
        network: Network;
      };
    }
  | {
      CoreCreate: {
        core: String;
        // both work somehow even though the action only takes CID in rust
        binary: CID | Uint8Array;
        tags: Tags;
      };
    }
  | {
      CoreRemove: {
        core: String;
      };
    }
  | {
      CoreChange: {
        core: String;
        state?: CID;
      };
    }
  | {
      CoreUpgrade: {
        core: String;

        /// The new binary.
        binary: CID;

        /// Migrate action.
        /// Must deserialize to a action using the new `binary`.
        migrate?: CID;
      };
    }
  | {
      CoreTagsInsert: {
        core: String;
        tags: Tags;
      };
    }
  | {
      CoreTagsRemove: {
        core: String;
        tags: Tags;
      };
    }
  | {
      GuardCreate: {
        guard: String;
        binary: CID;
        tags: Tags;
      };
    }
  | {
      GuardRemove: {
        guard: String;
      };
    }
  | {
      GuardUpgrade: {
        guard: String;
        /// The new binary.
        binary: CID;
      };
    }
  | {
      GuardTagsInsert: {
        guard: String;
        tags: Tags;
      };
    }
  | {
      GuardTagsRemove: {
        guard: String;
        tags: Tags;
      };
    };

File: co-uniffi/flutter/lib/src/cores/co.dart
import '../../co_flutter.dart';

/// See: ../../../../../cores/co/src/lib.rs
class CoreCo implements DagCborCodecProvider {
  final String /* CoId */ id;
  final List<dynamic>? /* Tags */ tags;
  final String /* String */ name;
  final Cid /* Cid */ binary;
  final CoMap<String, Participant> /* CoMap<Did, Participant> */ participants;
  final Map<String, Core> /* BTreeMap<String, Core> */ cores;
  final Map<String, Map<String, dynamic>>? /* BTreeMap<String, Guard> */ guards;
  final dynamic /* Option<Vec<Key>> */ keys;
  final dynamic /* CoSet<Network> */ network;

  CoreCo({
    required this.id,
    required this.tags,
    required this.name,
    required this.binary,
    required this.participants,
    required this.cores,
    required this.guards,
    required this.keys,
    required this.network,
  });

  static final codec = CoreCoDagCborCodec();

  @override
  DagCborCodec<dynamic> get dagCborCodec => codec;
}

class CoreCoDagCborCodec implements DagCborCodec<CoreCo> {
  @override
  CoreCo fromDagCborValue(data) {
    return CoreCo(
      id: data["id"],
      tags: data["t"],
      name: data["n"],
      binary: data["b"],
      participants: CoMap(data["p"], codec: Participant.codec),
      cores: (data["c"] as Map<dynamic, dynamic>).map((core_name, core) =>
          MapEntry(core_name as String, Core.codec.fromDagCborValue(core))),
      guards: data["g"],
      keys: data["k"],
      network: data["s"],
    );
  }

  @override
  dynamic toDagCborValue(CoreCo value) {
    return {
      "id": value.id,
      "t": value.tags,
      "n": value.name,
      "b": value.binary,
      "p": value.participants,
      "c": value.cores
          .map((core_name, core) => MapEntry(core_name, core.toDagCborValue())),
      "g": value.guards,
      "k": value.keys,
      "s": value.network,
    };
  }
}

class Core implements DagCborCodecProvider {
  final Cid /* Cid */ binary;
  final dynamic /* Tags */ tags;
  final Cid? /* Option<Cid> */ state;

  Core({required this.binary, required this.tags, required this.state});

  static final codec = CoreDagCborCodec();

  @override
  DagCborCodec<dynamic> get dagCborCodec => codec;
}

class CoreDagCborCodec implements DagCborCodec<Core> {
  @override
  Core fromDagCborValue(value) {
    return Core(
      binary: value["binary"],
      tags: value["tags"],
      state: value["state"],
    );
  }

  @override
  dynamic toDagCborValue(Core value) {
    return {
      "binary": value.binary,
      "tags": value.tags,
      "state": value.state,
    };
  }
}

class Participant implements DagCborCodecProvider {
  final dynamic /* Did */ did;
  final dynamic /* ParticipantState */ state;
  final dynamic /* Tags */ tags;

  Participant({required this.did, required this.state, required this.tags});

  static final codec = ParticipantDagCborCodec();

  @override
  DagCborCodec<dynamic> get dagCborCodec => codec;
}

class ParticipantDagCborCodec implements DagCborCodec<Participant> {
  @override
  Participant fromDagCborValue(value) {
    return Participant(
      did: value["did"],
      state: value["state"],
      tags: value["tags"],
    );
  }

  @override
  dynamic toDagCborValue(Participant value) {
    return {
      "did": value.did,
      "state": value.state,
      "tags": value.tags,
    };
  }
}
